### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by NataeSz**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -131,4 +131,5 @@ dmypy.json
 # own
 *.png
 *.csv
-*.idea
\ No newline at end of file
+*.idea
+*.gif
```
Comment (prybicki) [inline .gitignore]: Missing end of line: using `cat` to display file will look ugly.

File: README.md [hunk 2]
```diff
@@ -3,13 +3,32 @@
 Genetic algorithm for a simple rule-based robot on a 2D-grid
 
 ## Overview
+
+Genetic algorithm is search heuristic that reflects the process of natural selection.
```
Comment (prybicki) [inline README.md]:
```suggestion
Genetic algorithm is a heuristic search that reflects the process of natural selection.
```

File: README.md [hunk 3]
```diff
@@ -3,13 +3,32 @@
...
+It consists of five phases:
+- Initial population
+- Fitness function
+- Selection
+- Crossover
+- Mutation
```
Comment (prybicki) [inline README.md]: These are not phases ;)

File: README.md [hunk 4]
```diff
@@ -3,13 +3,32 @@
...
+
+In this project, the genetic algorithm has been used to find the best set of movement rules 
+for a robot on a grid that collects cans.
```
Comment (prybicki) [inline README.md]: No need to break line.

Comment (prybicki) [reply README.md]: Below, you will find an examplary behavior of a robot from 1000th generation. It's not optimal, but it's definitely better than the random one.

File: README.md [hunk 5]
```diff
@@ -3,13 +3,32 @@
...
+![Robby tracking](./docs/Robby_tracking.gif)
+
+## Details
 ### Robby and his World
 
 Robby is a very simple robot. He lives in a simulated world, consisting of a small (10x10) grid of squares.
-His life goal is to collect empty soda cans, which are distributed randomly in the world. Each square of the grid can contain zero or one empty soda can. Robby has very bad vision. He only sees the contents of the square he stands in and the four side-adjacent cells (a square can be empty, wall or contain empty soda can). Based on the input from 5 cells, in each simulation step, he can choose from one of the following actions:
+His life goal is to collect empty soda cans, which are distributed randomly in the world. 
+Each square of the grid can contain zero or one empty soda can. Robby has very bad vision. 
+He only sees the contents of the square he stands in and the four side-adjacent cells
```
Comment (prybicki) [inline README.md]:
```suggestion
He only sees the content of the square he stands in and the four side-adjacent cells 
```

File: README.md [hunk 6]
```diff
@@ -3,13 +3,32 @@
...
+
+## Details
 ### Robby and his World
 
 Robby is a very simple robot. He lives in a simulated world, consisting of a small (10x10) grid of squares.
-His life goal is to collect empty soda cans, which are distributed randomly in the world. Each square of the grid can contain zero or one empty soda can. Robby has very bad vision. He only sees the contents of the square he stands in and the four side-adjacent cells (a square can be empty, wall or contain empty soda can). Based on the input from 5 cells, in each simulation step, he can choose from one of the following actions:
+His life goal is to collect empty soda cans, which are distributed randomly in the world. 
+Each square of the grid can contain zero or one empty soda can. Robby has very bad vision. 
+He only sees the contents of the square he stands in and the four side-adjacent cells 
+(a square can be empty, wall or contain empty soda can). Based on the input from 5 cells, in each simulation step,
```
Comment (prybicki) [inline README.md]:
```suggestion
(a square can be empty, be a wall, or contain an empty soda can). Based on the input from these 5 cells, in each simulation step, 
```

File: README.md [hunk 7]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
 
 Every generation (except the first one) are created from the previous one
 by sampling two parents and producing a child until N children are created.
-For a robot, the probability of being chosen to be a parent should depend on its score (better score = higher probability).
-Child genetic code is created by concatenating a prefix from one parent and a suffix from the other one, to create a complete rule table.
+For a robot, the probability of being chosen to be a parent should depend on its score 
+(better score = higher probability).
+Child genetic code is created by concatenating a prefix from one parent and a suffix from the other one,
```
Comment (prybicki) [inline README.md]:
```suggestion
The child's genetic code is created by concatenating a prefix from one parent and a suffix from the other one, 
```

File: README.md [hunk 8]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
 
 Every generation (except the first one) are created from the previous one
 by sampling two parents and producing a child until N children are created.
-For a robot, the probability of being chosen to be a parent should depend on its score (better score = higher probability).
-Child genetic code is created by concatenating a prefix from one parent and a suffix from the other one, to create a complete rule table.
+For a robot, the probability of being chosen to be a parent should depend on its score 
+(better score = higher probability).
+Child genetic code is created by concatenating a prefix from one parent and a suffix from the other one, 
+to create a complete rule table.
 Additionally, a small number of mutations should be introduced.
```
Comment (prybicki) [inline README.md]:
```suggestion
Additionally, a small number of mutations are introduced.
```

File: README.md [hunk 9]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
...
 
 
+## Results 
+
+Sample results of the evolution efficiency are illustrated by the following charts 
+of the average and maximal score of robots in each generation.
+
+![Avg fitness](./docs/avg_fitness.png)
+The evolution consists of periods of constant fitness (no progress in the following generations)
```
Comment (prybicki) [inline README.md]:
```suggestion
The evolution consists of periods of constant fitness (no progress in thousands of generations) 
```

File: README.md [hunk 10]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
...
 
 
+## Results 
+
+Sample results of the evolution efficiency are illustrated by the following charts 
+of the average and maximal score of robots in each generation.
+
+![Avg fitness](./docs/avg_fitness.png)
+The evolution consists of periods of constant fitness (no progress in the following generations) 
+and periods of substantial innovations represented by rapid growths of fitness,
```
Comment (prybicki) [inline README.md]:
```suggestion
and periods of substantial innovation represented by a rapid growth of fitness, 
```

File: README.md [hunk 11]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
...
+
+Sample results of the evolution efficiency are illustrated by the following charts 
+of the average and maximal score of robots in each generation.
+
+![Avg fitness](./docs/avg_fitness.png)
+The evolution consists of periods of constant fitness (no progress in the following generations) 
+and periods of substantial innovations represented by rapid growths of fitness, 
+similarly as in a biological evolution. <br>
+The average score rises shortly after the evolution starts.
+During the evolution, trait function growths reflects biological exaptations
```
Comment (prybicki) [inline README.md]: Unreadable.

File: README.md [hunk 12]
```diff
@@ -49,9 +68,34 @@ After all individuals have been tested, a new generation is created.
...
+Sample results of the evolution efficiency are illustrated by the following charts 
+of the average and maximal score of robots in each generation.
+
+![Avg fitness](./docs/avg_fitness.png)
+The evolution consists of periods of constant fitness (no progress in the following generations) 
+and periods of substantial innovations represented by rapid growths of fitness, 
+similarly as in a biological evolution. <br>
+The average score rises shortly after the evolution starts.
+During the evolution, trait function growths reflects biological exaptations 
+and intermingle with the groups of constant, non-adaptive generations.
```
Comment (prybicki) [inline README.md]: the?

File: src/plot.py [hunk 13]
```diff
@@ -63,6 +63,10 @@ def plot_movements(genetic_code: List[int] = None, grid: Grid = None, grid_from_
...
         total_points += points_for_action
-        print()
-        print('Points for the action:\t', points_for_action)
+        print('\nPoints for the action:\t', points_for_action)
         print('Total points:\t\t', total_points)
+        print('\nLegend:')
+        print('\t', MovementRepr.current_position_repr[MovementRepr.can_repr[0]], '\t- robot on an empty square')
+        print('\t', MovementRepr.current_position_repr[MovementRepr.can_repr[1]], '\t- robot on a square with a can')
+        print('\t', MovementRepr.can_repr[0], '\t- empty square')
+        print('\t', MovementRepr.can_repr[1], '\t- can')
```
Comment (prybicki) [inline src/plot.py]: An f-string would make it much more clear.

**Review by prybicki — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (NataeSz) [reply README.md]: Deleted; more details to the definitions added

File: README.md (hunk 11 above)
Comment (NataeSz) [reply README.md]: Simplified in the next commit

File: src/plot.py (hunk 13 above)
Comment (NataeSz) [reply src/plot.py]: Done :)

**Review by prybicki — APPROVED** [review]

**Pull request merged**